### PR TITLE
Try wasmtime

### PR DESCRIPTION
**Pull request by gterzian**

### CODE REVIEW

**Comment by gterzian** [comment]
@damip @yvan-sraka @AureliaDolo @damip Ok so this is a simple example of using wasmtime, to :

1. Call a module,
2. Call another module from the first one,
3. Mutate some shared state.

Please let me know what you think.

File: execution/src/vm.rs [hunk 1]
```diff
@@ -98,48 +38,141 @@ pub struct ExecutionContext {
...
+            let mut modules: AddressHashMap<Module> = Default::default();
+            let wat = r#"
+            (module
+                (import "host" "set" (func $host_set (param i32)))
+                (import "host" "call" (func $host_call (param i32 i32)))
+                (func (export "main")
+                    i32.const 3
+                    call $host_set)
+            )
+        "#;
```
Comment (adrien-zinger) [inline execution/src/vm.rs:66]: where this code come from? the style of arguments pointer+size isn't standardized if you want to develop in c++, grain etc.

File: execution/src/vm.rs [hunk 2]
```diff
@@ -98,48 +38,141 @@ pub struct ExecutionContext {
...
+            move |mut caller: Caller<'_, SharedExecutionContext>, ptr: i32, len: i32| {
+                println!("Start host call");
+                let mem = match caller.get_export("memory") {
+                    Some(Extern::Memory(mem)) => mem,
+                    _ => return Err(Trap::new("failed to find host memory")),
+                };
+                let data = mem
+                    .data(&caller)
+                    .get(ptr as u32 as usize..)
+                    .and_then(|arr| arr.get(..len as u32 as usize));
```
Comment (adrien-zinger) [inline execution/src/vm.rs:88]: that is a nice feature

File: execution/src/vm.rs [hunk 3]
```diff
@@ -98,48 +30,138 @@ pub struct ExecutionContext {
...
+                    (module, imports)
+                };
+
+                // Instantiate, and call, the module.
+                let instance = Instance::new(&mut caller, &module, &imports).unwrap();
+                let foo = instance
+                    .get_typed_func::<(), (), _>(&mut caller, "main")
+                    .unwrap();
+                foo.call(&mut caller, ()).unwrap();
+                Ok(())
```
Comment (adrien-zinger) [inline execution/src/vm.rs:116]: Ok now we need to pass a value between instances and we're good

File: execution/src/vm.rs (hunk 1 above)
Comment (gterzian) [reply execution/src/vm.rs:66]: I copied it from various examples found in the docs: https://docs.rs/wasmtime/latest/wasmtime/

**Comment by AureliaDolo** [comment]
@gterzian are you still working on that ? We can either close or merge that issue

**Comment by yvan-sraka** [comment]
We can merge it under a feature flag! ⛳️

**Comment by gterzian** [comment]
I think it's best to close at this point, since it was just a proof of concept, I would have to restructure the code to make it mergeable.

**Pull request closed without merging**